README
WWW-Wordpress

The README is used to introduce the module and provide instructions on
how to install the module, any machine dependencies it may have (for
example C compilers and installed libraries) and any other information
that should be provided before the module is installed.

A README file is required for CPAN modules since CPAN extracts the README
file from a module distribution so that people browsing the archive
can use it to get an idea of the module's uses. It is usually a good idea
to provide version information here so that people can decide whether
fixes for the module are worth downloading.


INSTALLATION
METHOD 1 (RECOMMENDED)

$ sudo cpanm .

METHOD 2
To install this module, run the following commands:

	perl Makefile.PL
	make
	make test
	make install

SUPPORT AND DOCUMENTATION

After installing, you can find documentation for this module with the
perldoc command.

    perldoc WWW::Wordpress

You can also look for information at:

    RT, CPAN's request tracker (report bugs here)
        http://rt.cpan.org/NoAuth/Bugs.html?Dist=WWW-Wordpress

    AnnoCPAN, Annotated CPAN documentation
        http://annocpan.org/dist/WWW-Wordpress

    CPAN Ratings
        http://cpanratings.perl.org/d/WWW-Wordpress

    Search CPAN
        http://search.cpan.org/dist/WWW-Wordpress/


LICENSE AND COPYRIGHT

Copyright (C) 2014 Daniel Torres

This program is free software; you can redistribute it and/or modify it
under the terms of the the Artistic License (2.0). You may obtain a
copy of the full license at:

L<http://www.perlfoundation.org/artistic_license_2_0>

Any use, modification, and distribution of the Standard or Modified
Versions is governed by this Artistic License. By using, modifying or
distributing the Package, you accept this license. Do not use, modify,
or distribute the Package, if you do not accept this license.

If your Modified Version has been derived from a Modified Version made
by someone other than you, you are nevertheless required to ensure that
your Modified Version complies with the requirements of this license.

This license does not grant you the right to use any trademark, service
mark, tradename, or logo of the Copyright Holder.

This license includes the non-exclusive, worldwide, free-of-charge
patent license to make, have made, use, offer to sell, sell, import and
otherwise transfer the Package with respect to any patent claims
licensable by the Copyright Holder that are necessarily infringed by the
Package. If you institute patent litigation (including a cross-claim or
counterclaim) against any party alleging that the Package constitutes
direct or contributory patent infringement, then this Artistic License
to you shall terminate on the date that such litigation is filed.

Disclaimer of Warranty: THE PACKAGE IS PROVIDED BY THE COPYRIGHT HOLDER
AND CONTRIBUTORS "AS IS' AND WITHOUT ANY EXPRESS OR IMPLIED WARRANTIES.
THE IMPLIED WARRANTIES OF MERCHANTABILITY, FITNESS FOR A PARTICULAR
PURPOSE, OR NON-INFRINGEMENT ARE DISCLAIMED TO THE EXTENT PERMITTED BY
YOUR LOCAL LAW. UNLESS REQUIRED BY LAW, NO COPYRIGHT HOLDER OR
CONTRIBUTOR WILL BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, OR
CONSEQUENTIAL DAMAGES ARISING IN ANY WAY OUT OF THE USE OF THE PACKAGE,
EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.



Changes
Revision history for WWW-Wordpress

0.01    Date/time
        First version, released on an unsuspecting world.



lib/WWW/Wordpress.pm
# Copyleft 2014 Daniel Torres 
# daniel.torres at owasp.org
# All rights released.
package WWW::Wordpress;


use strict;
use warnings FATAL => 'all';
use Carp qw(croak);
use Moose;
use Net::SSL (); # From Crypt-SSLeay
use LWP::UserAgent;
use HTTP::Cookies;
use JSON qw( decode_json ); 
use URI;
use Data::Dumper;

our $VERSION = '1.3';

###### default values #####
use constant WORDPRESS_URL => 'https://public-api.Wordpress.com/rest/v1/';
$Net::HTTPS::SSL_SOCKET_CLASS = "Net::SSL"; # Force use of Net::SSL for proxy compatibility
##############################

{
has 'blog', is => 'rw', isa => 'Str',required => 1;	
has 'blog_id', is => 'rw', isa => 'Str',required => 1;	
has 'access_token', is => 'rw', isa => 'Str',required => 1;

has proxy_host      => ( isa => 'Str', is => 'rw', default => '' );
has proxy_port      => ( isa => 'Str', is => 'rw', default => '' );
has proxy_user      => ( isa => 'Str', is => 'rw', default => '' );
has proxy_pass      => ( isa => 'Str', is => 'rw', default => '' );
has proxy_env      => ( isa => 'Str', is => 'rw', default => '' );

has browser  => ( isa => 'Object', is => 'rw', lazy => 1, builder => '_build_browser' );


##### site info #######       
sub site_info
{
my $self = shift;
my %options = @_;
my $blog = $self->blog;
my $response = $self->dispatch(path => "sites/$blog",method => 'GET');
#my content = $response->content;

my $response_json = decode_json($response->content);
#print Dumper $response_json;
my $post_count = $response_json->{'post_count'};
my $subscribers_count = $response_json->{'subscribers_count'};
return {post_count => $post_count, subscribers_count => $subscribers_count};
}

###################################### blog  functions ###################


# post to blog
sub post
{
my $self = shift;
my %options = @_;
my $post_data = $options{ post_data };
my $blog = $self->blog;

my $response = $self->dispatch(path => "sites/$blog/posts/new",method => 'POST',post_data =>$post_data);
#print "\n \n".$response->content;
my $response_json = decode_json($response->content);
my $id = $response_json->{'ID'};
my $URL = $response_json->{'URL'};

return {id => $id, URL => $URL};
}    


#edit post
sub edit_post
{
my $self = shift;
my %options = @_;
my $post_data = $options{ post_data };
my $post_id = $options{ post_id };
#print Dumper %options;
my $blog = $self->blog;
my $response = $self->dispatch(path => "sites/$blog/posts/$post_id",method => 'POST',post_data =>$post_data);
my $response_json = decode_json($response->content);
#print Dumper $response_json;
my $status = $response_json->{'status'};
return $status ;
}


###################################### Users  functions ##################

# Follow a blog
sub follow
{
my $self = shift;
my %options = @_;
my $blog = $options{ blog };
my $post_data = $options{ post_data };

my $response = $self->dispatch(path => "sites/$blog/follows/new",method => 'POST',post_data =>$post_data);
my $response_json = decode_json($response->content);
#print Dumper $response_json;
my $status = $response_json->{'success'};
return $status ;
}

# unfollow a blog
sub unfollow
{
my $self = shift;
my %options = @_;
my $blog = $options{ blog };
my $post_data = $options{ post_data };
my $response = $self->dispatch(path => "sites/$blog/follows/mine/delete",method => 'POST',post_data =>$post_data);
my $response_json = decode_json($response->content);
#print Dumper $response_json;
my $status = $response_json->{'success'};
#my $is_following = $response_json->{'is_following'};
return $status ;
}


###################################### internal functions ###################
sub dispatch {    
my $self = shift;
my %options = @_;
my $access_token = $self->access_token;

$self->browser->default_header('Authorization' => "Bearer $access_token");

my $base = WORDPRESS_URL;
my $path = $options{ path };
my $method = $options{ method };

my $url = $base.$path;
#print "url $url \n";
my $response = '';
if ($method eq 'GET')
  { $response = $self->browser->get($url);}
  
if ($method eq 'POST')
  {     
   my $post_data = $options{ post_data };        
   $response = $self->browser->post($url,$post_data);
  }  
  
return $response;
}

sub _build_browser {    
my $self = shift;

my $proxy_host = $self->proxy_host;
my $proxy_port = $self->proxy_port;
my $proxy_user = $self->proxy_user;
my $proxy_pass = $self->proxy_pass;
my $proxy_env = $self->proxy_env;

my $browser = LWP::UserAgent->new;
$browser->timeout(20);
$browser->show_progress(1);
#print "proxy_env $proxy_env \n";

if ( $proxy_env eq 'ENV' )
{
$Net::HTTPS::SSL_SOCKET_CLASS = "Net::SSL"; # Force use of Net::SSL
$ENV{HTTPS_PROXY} = "http://".$proxy_host.":".$proxy_port;
}
else
{
  if (($proxy_user ne "") && ($proxy_host ne ""))
  {
   $browser->proxy(['http', 'https'], 'http://'.$proxy_user.':'.$proxy_pass.'@'.$proxy_host.':'.$proxy_port); # Using a private proxy
  }
  elsif ($proxy_host ne "")
    { $browser->proxy(['http', 'https'], 'http://'.$proxy_host.':'.$proxy_port);} # Using a public proxy
  else
    { $browser->env_proxy;} # No proxy       
}     
    
return $browser;
}

}

1;



__END__

=head1 NAME

WWW::Wordpress - Wordpress API interface .


=head1 SYNOPSIS


Usage:

   use WWW::Wordpress;
   $ENV{PERL_LWP_SSL_VERIFY_HOSTNAME} = 0;
   my $wordpress = WWW::Wordpress->new( blog => 'blog.Wordpress.com',
					blog_id => '00000',
					access_token => 'XXXXX');					



=head1 DESCRIPTION

Wordpress API interface

=head1 FUNCTIONS

=head2 constructor

    my $wordpress = WWW::Wordpress->new( blog => 'blog.Wordpress.com',
					blog_id => '00000',
					access_token => 'XXXXX');	

To get your Access Token check

http://developer.Wordpress.com/docs/oauth2/#receiving-an-access-token

=head2 site_info

    $site_info = $wordpress->site_info;
    $post_count = $site_info->{post_count};
    $subscribers_count = $site_info->{subscribers_count};
    print "post_count $post_count \n";
    print "subscribers_count $subscribers_count \n";

stats of the site.

=head2 post

    $content = "this is my test body ";
    $title = "new title";
    $tags = "linux";
    my $post_data = { content => $content,title => $title,tags => $tags };
    $post_info = $wordpress->post(post_data => $post_data );
    $id = $post_info->{id};
    $URL = $post_info->{URL};
    print "id $id URL $URL  \n";


Post an article

=head2 edit_post

    $new_content = "Edited test";
    $post_id = 19;

    my $post_data = { content => $new_content};
    $status = $wordpress->edit_post(post_data => $post_data, post_id => $post_id);
    print "status $status \n";

Edit a post by post id

=head2 follow

    my $post_data = { pretty => 1};
    $status = $wordpress->follow(blog => 'blog.Wordpress.com',post_data => $post_data);  
    print "status $status \n";

Follow another blog
   
=head2 unfollow

    my $post_data = { pretty => 1};
    $status = $wordpress->unfollow(blog => 'blog.Wordpress.com',post_data => $post_data);  
    print "status $status \n";

Unfollow blog
   
=head2 dispatch

 Internal function         
                  
=head1 LICENSE

This program is free software; you can redistribute it and/or modify it under the same terms as Perl itself.

See http://www.perl.com/perl/misc/Artistic.html
=cut


lib/WWW/wordpress.pm
# Copyleft 2014 Daniel Torres 
# daniel.torres at owasp.org
# All rights released.
package WWW::Wordpress;


use strict;
use warnings FATAL => 'all';
use Carp qw(croak);
use Moose;
use Net::SSL (); # From Crypt-SSLeay
use LWP::UserAgent;
use HTTP::Cookies;
use JSON qw( decode_json ); 
use URI;
use Data::Dumper;

our $VERSION = '1.2';

###### default values #####
use constant WORDPRESS_URL => 'https://public-api.wordpress.com/rest/v1/';
$Net::HTTPS::SSL_SOCKET_CLASS = "Net::SSL"; # Force use of Net::SSL for proxy compatibility
##############################

{
has 'blog', is => 'rw', isa => 'Str',required => 1;	
has 'blog_id', is => 'rw', isa => 'Str',required => 1;	
has 'access_token', is => 'rw', isa => 'Str',required => 1;

has proxy_host      => ( isa => 'Str', is => 'rw', default => '' );
has proxy_port      => ( isa => 'Str', is => 'rw', default => '' );
has proxy_user      => ( isa => 'Str', is => 'rw', default => '' );
has proxy_pass      => ( isa => 'Str', is => 'rw', default => '' );
has proxy_env      => ( isa => 'Str', is => 'rw', default => '' );

has browser  => ( isa => 'Object', is => 'rw', lazy => 1, builder => '_build_browser' );


##### site info #######       
sub site_info
{
my $self = shift;
my %options = @_;
my $blog = $self->blog;
my $response = $self->dispatch(path => "sites/$blog",method => 'GET');
#my content = $response->content;

my $response_json = decode_json($response->content);
#print Dumper $response_json;
my $post_count = $response_json->{'post_count'};
my $subscribers_count = $response_json->{'subscribers_count'};
return {post_count => $post_count, subscribers_count => $subscribers_count};
}

###################################### blog  functions ###################


# post to blog
sub post
{
my $self = shift;
my %options = @_;
my $post_data = $options{ post_data };
my $blog = $self->blog;

my $response = $self->dispatch(path => "sites/$blog/posts/new",method => 'POST',post_data =>$post_data);
#print "\n \n".$response->content;
my $response_json = decode_json($response->content);
my $id = $response_json->{'ID'};
my $URL = $response_json->{'URL'};

return {id => $id, URL => $URL};
}    


#edit post
sub edit_post
{
my $self = shift;
my %options = @_;
my $post_data = $options{ post_data };
my $post_id = $options{ post_id };
#print Dumper %options;
my $blog = $self->blog;
my $response = $self->dispatch(path => "sites/$blog/posts/$post_id",method => 'POST',post_data =>$post_data);
my $response_json = decode_json($response->content);
#print Dumper $response_json;
my $status = $response_json->{'status'};
return $status ;
}


###################################### Users  functions ##################

# Follow a blog
sub follow
{
my $self = shift;
my %options = @_;
my $blog = $options{ blog };
my $post_data = $options{ post_data };

my $response = $self->dispatch(path => "sites/$blog/follows/new",method => 'POST',post_data =>$post_data);
my $response_json = decode_json($response->content);
#print Dumper $response_json;
my $status = $response_json->{'success'};
return $status ;
}

# unfollow a blog
sub unfollow
{
my $self = shift;
my %options = @_;
my $blog = $options{ blog };
my $post_data = $options{ post_data };
my $response = $self->dispatch(path => "sites/$blog/follows/mine/delete",method => 'POST',post_data =>$post_data);
my $response_json = decode_json($response->content);
#print Dumper $response_json;
my $status = $response_json->{'success'};
#my $is_following = $response_json->{'is_following'};
return $status ;
}


###################################### internal functions ###################
sub dispatch {    
my $self = shift;
my %options = @_;
my $access_token = $self->access_token;

$self->browser->default_header('Authorization' => "Bearer $access_token");

my $base = WORDPRESS_URL;
my $path = $options{ path };
my $method = $options{ method };

my $url = $base.$path;
#print "url $url \n";
my $response = '';
if ($method eq 'GET')
  { $response = $self->browser->get($url);}
  
if ($method eq 'POST')
  {     
   my $post_data = $options{ post_data };        
   $response = $self->browser->post($url,$post_data);
  }  
  
return $response;
}

sub _build_browser {    
my $self = shift;

my $proxy_host = $self->proxy_host;
my $proxy_port = $self->proxy_port;
my $proxy_user = $self->proxy_user;
my $proxy_pass = $self->proxy_pass;
my $proxy_env = $self->proxy_env;

my $browser = LWP::UserAgent->new;
$browser->timeout(20);
$browser->show_progress(1);
#print "proxy_env $proxy_env \n";

if ( $proxy_env eq 'ENV' )
{
$Net::HTTPS::SSL_SOCKET_CLASS = "Net::SSL"; # Force use of Net::SSL
$ENV{HTTPS_PROXY} = "http://".$proxy_host.":".$proxy_port;
}
else
{
  if (($proxy_user ne "") && ($proxy_host ne ""))
  {
   $browser->proxy(['http', 'https'], 'http://'.$proxy_user.':'.$proxy_pass.'@'.$proxy_host.':'.$proxy_port); # Using a private proxy
  }
  elsif ($proxy_host ne "")
    { $browser->proxy(['http', 'https'], 'http://'.$proxy_host.':'.$proxy_port);} # Using a public proxy
  else
    { $browser->env_proxy;} # No proxy       
}     
    
return $browser;
}

}

1;



__END__

=head1 NAME

WWW::Wordpress - wordpress API interface .


=head1 SYNOPSIS


Usage:

   use WWW::Wordpress;
   $ENV{PERL_LWP_SSL_VERIFY_HOSTNAME} = 0;
   my $wordpress = WWW::Wordpress->new( blog => 'blog.wordpress.com',
					blog_id => '00000',
					access_token => 'XXXXX');					



=head1 DESCRIPTION

wordpress API interface

=head1 FUNCTIONS

=head2 constructor

    my $wordpress = WWW::Wordpress->new( blog => 'blog.wordpress.com',
					blog_id => '00000',
					access_token => 'XXXXX');	

To get your Access Token check

http://developer.wordpress.com/docs/oauth2/#receiving-an-access-token

=head2 site_info

    $site_info = $wordpress->site_info;
    $post_count = $site_info->{post_count};
    $subscribers_count = $site_info->{subscribers_count};
    print "post_count $post_count \n";
    print "subscribers_count $subscribers_count \n";

stats of the site.

=head2 post

    $content = "this is my test body ";
    $title = "new title";
    $tags = "linux";
    my $post_data = { content => $content,title => $title,tags => $tags };
    $post_info = $wordpress->post(post_data => $post_data );
    $id = $post_info->{id};
    $URL = $post_info->{URL};
    print "id $id URL $URL  \n";


Post an article

=head2 edit_post

    $new_content = "Edited test";
    $post_id = 19;

    my $post_data = { content => $new_content};
    $status = $wordpress->edit_post(post_data => $post_data, post_id => $post_id);
    print "status $status \n";

Edit a post by post id

=head2 follow

    my $post_data = { pretty => 1};
    $status = $wordpress->follow(blog => 'blog.wordpress.com',post_data => $post_data);  
    print "status $status \n";

Follow another blog
   
=head2 unfollow

    my $post_data = { pretty => 1};
    $status = $wordpress->unfollow(blog => 'blog.wordpress.com',post_data => $post_data);  
    print "status $status \n";

Unfollow blog
   
=head2 dispatch

 Internal function         
                  
=head1 LICENSE

This program is free software; you can redistribute it and/or modify it under the same terms as Perl itself.

See http://www.perl.com/perl/misc/Artistic.html
=cut
